# Lab 2: Data preperation

Following on from EDA last week, this notebook starts the process of data preprocessing or data preporation so it is in the best state for machine learning.

# A: Import Python Modules and CSV Dataset

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import RobustScaler
import matplotlib.pyplot as plt
%matplotlib inline

df = pd.read_csv("./bank.csv")
df.head()

age          job   marital  education default  balance housing loan  \
0  32.0   technician    single   tertiary      no      392     yes   no   
1  39.0   technician  divorced  secondary      no      688     yes  yes   
2  59.0      retired   married  secondary      no     1035     yes  yes   
3  47.0  blue-collar   married  secondary      no      398     yes  yes   
4  54.0      retired   married  secondary      no     1004     yes   no   

    contact  day month  duration  campaign  pdays  previous poutcome  \
0  cellular    1   apr       957         2    131         2  failure   
1  cellular    1   apr       233         2    133         1  failure   
2  cellular    1   apr       126         2    239         1  failure   
3  cellular    1   apr       274         1    238         2  failure   
4  cellular    1   apr       479         1    307         1  failure   

  subscribed  
0         no  
1         no  
2         no  
3         no  
4         no


To see what we are working with, we can then detect and sum the number of missing values:

In [2]:
df.isna().sum()

age            12
job            10
marital         0
education     104
default         0
balance         0
housing         0
loan            0
contact       191
day             0
month           0
duration        0
campaign        0
pdays           0
previous        0
poutcome      454
subscribed      0
dtype: int64

This shows 5 columns to have missing data with poutcome having the most (454)


Make a copy of the dataset to work on:

In [3]:
df_original = df


# B: Encode All Categorical Attributes as Numerical

The following are are categorical:

job, marital, education, contact, poutcome

In [4]:
# Create mappers

## marital mapper

marital_scale_mapper = {
    "single": 1,
    "married": 2,
    "divorced": 3
}

df['marital'] = df['marital'].replace(marital_scale_mapper)


## education mapper

education_scale_mapper = {
    "primary": 1,
    "secondary": 2,
    "tertiary": 3
}

df['education'] = df['education'].replace(education_scale_mapper)


## contact mapper

contact_scale_mapper = {
    "cellular": 1,
    "telephone": 2
}

df['contact'] = df['contact'].replace(contact_scale_mapper)


## month

month_scale_mapper = {
    "jan": 1,
    "feb": 2,
    "mar": 3,
    "apr": 4,
    "may": 5,
    "jun": 6,
    "jul": 7,
    "aug": 8,
    "sep": 9,
    "oct": 10,
    "nov": 11,
    "dec": 12,
}
df['month'] = df['month'].replace(month_scale_mapper)


## poutcome mapper

poutcome_scale_mapper = {
    "success": 1,
    "failure": 2,
    "other": 3
}

df['poutcome'] = df['poutcome'].replace(poutcome_scale_mapper)

df.tail()

age           job  marital  education default  balance housing loan  \
1995  20.0       student        1        NaN      no     2785      no   no   
1996  28.0        admin.        1        2.0      no      127      no   no   
1997  81.0       retired        2        1.0      no     1154      no   no   
1998  46.0      services        2        1.0      no     4343     yes   no   
1999  40.0  entrepreneur        2        2.0      no     6403      no   no   

      contact  day  month  duration  campaign  pdays  previous  poutcome  \
1995      1.0   16      9       327         2     -1         0       NaN   
1996      1.0   16      9      1334         2     -1         0       NaN   
1997      2.0   17      9       231         1     -1         0       NaN   
1998      NaN   20      9       185         1     -1         0       NaN   
1999      1.0   22      9       208         2     -1         0       NaN   

     subscribed  
1995        yes  
1996        yes  
1997        yes  
1998        yes  
1999        yes


# C: One-hot encode all other categorical attributes:

In [5]:
# first one-hot encode the categorical columns with NaNs

df = pd.get_dummies(df, columns=['job'], 
                        dummy_na=True, 
                        drop_first=True)
df.tail()

age  marital  education default  balance housing loan  contact  day  \
1995  20.0        1        NaN      no     2785      no   no      1.0   16   
1996  28.0        1        2.0      no      127      no   no      1.0   16   
1997  81.0        2        1.0      no     1154      no   no      2.0   17   
1998  46.0        2        1.0      no     4343     yes   no      NaN   20   
1999  40.0        2        2.0      no     6403      no   no      1.0   22   

      month  ...  job_entrepreneur  job_housemaid  job_management  \
1995      9  ...                 0              0               0   
1996      9  ...                 0              0               0   
1997      9  ...                 0              0               0   
1998      9  ...                 0              0               0   
1999      9  ...                 1              0               0   

      job_retired  job_self-employed job_services  job_student  \
1995            0                  0            0            1   
1996            0                  0            0            0   
1997            1                  0            0            0   
1998            0                  0            1            0   
1999            0                  0            0            0   

      job_technician  job_unemployed  job_nan  
1995               0               0        0  
1996               0               0        0  
1997               0               0        0  
1998               0               0        0  
1999               0               0        0  

[5 rows x 27 columns]

In [6]:
# then one-hot encode all other categorical columns

df = pd.get_dummies(df, drop_first=True)
df.head()

age  marital  education  balance  contact  day  month  duration  campaign  \
0  32.0        1        3.0      392      1.0    1      4       957         2   
1  39.0        3        2.0      688      1.0    1      4       233         2   
2  59.0        2        2.0     1035      1.0    1      4       126         2   
3  47.0        2        2.0      398      1.0    1      4       274         1   
4  54.0        2        2.0     1004      1.0    1      4       479         1   

   pdays  ...  job_self-employed  job_services  job_student  job_technician  \
0    131  ...                  0             0            0               1   
1    133  ...                  0             0            0               1   
2    239  ...                  0             0            0               0   
3    238  ...                  0             0            0               0   
4    307  ...                  0             0            0               0   

   job_unemployed  job_nan  default_yes  housing_yes  loan_yes  subscribed_yes  
0               0        0            0            1         0               0  
1               0        0            0            1         1               0  
2               0        0            0            1         1               0  
3               0        0            0            1         1               0  
4               0        0            0            1         0               0  

[5 rows x 27 columns]


# D: Fill in all missing values

In [7]:
# How many  nulls/NaNs are in the updated dataset

df.isna().sum()

age                   12
marital                0
education            104
balance                0
contact              191
day                    0
month                  0
duration               0
campaign               0
pdays                  0
previous               0
poutcome             454
job_blue-collar        0
job_entrepreneur       0
job_housemaid          0
job_management         0
job_retired            0
job_self-employed      0
job_services           0
job_student            0
job_technician         0
job_unemployed         0
job_nan                0
default_yes            0
housing_yes            0
loan_yes               0
subscribed_yes         0
dtype: int64

In [8]:
# Fill in all missing values with the median of each attribute that is missing values

df['age'].fillna(df['age'].median(), inplace=True)
df['education'].fillna(df['education'].median(), inplace=True)
df['contact'].fillna(df['contact'].median(), inplace=True)
df['poutcome'].fillna(df['poutcome'].median(), inplace=True)

In [9]:
# Verify changes

df.isna().sum()

age                  0
marital              0
education            0
balance              0
contact              0
day                  0
month                0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
job_blue-collar      0
job_entrepreneur     0
job_housemaid        0
job_management       0
job_retired          0
job_self-employed    0
job_services         0
job_student          0
job_technician       0
job_unemployed       0
job_nan              0
default_yes          0
housing_yes          0
loan_yes             0
subscribed_yes       0
dtype: int64

In [10]:
df.tail()

age  marital  education  balance  contact  day  month  duration  \
1995  20.0        1        2.0     2785      1.0   16      9       327   
1996  28.0        1        2.0      127      1.0   16      9      1334   
1997  81.0        2        1.0     1154      2.0   17      9       231   
1998  46.0        2        1.0     4343      1.0   20      9       185   
1999  40.0        2        2.0     6403      1.0   22      9       208   

      campaign  pdays  ...  job_self-employed  job_services  job_student  \
1995         2     -1  ...                  0             0            1   
1996         2     -1  ...                  0             0            0   
1997         1     -1  ...                  0             0            0   
1998         1     -1  ...                  0             1            0   
1999         2     -1  ...                  0             0            0   

      job_technician  job_unemployed  job_nan  default_yes  housing_yes  \
1995               0               0        0            0            0   
1996               0               0        0            0            0   
1997               0               0        0            0            0   
1998               0               0        0            0            1   
1999               0               0        0            0            0   

      loan_yes  subscribed_yes  
1995         0               1  
1996         0               1  
1997         0               1  
1998         0               1  
1999         0               1  

[5 rows x 27 columns]

# E: Apply a scalar to rescale features